contrib/tracking.go

```diff
@@ -0,0 +1,256 @@
+package contrib
+
+/*
+#include "tracking.h"
+*/
+import "C"
+import (
+	"image"
+
+	"gocv.io/x/gocv"
+)
+
+// Tracker - the base interface for object tracking
+//
+// s.a: https://docs.opencv.org/master/d0/d0a/classcv_1_1Tracker.html
+type Tracker interface {
+	// Init initializes the tracker with a known bounding box that surrounded the target
+	// note: this can only be called once. if you loose the object, you have to Close() the instance,
+	// create a new one, and call Init on it again
+	//
+	// https://docs.opencv.org/master/d0/d0a/classcv_1_1Tracker.html#a4d285747589b1bdd16d2e4f00c3255dc
+	Init(image gocv.Mat, boundingBox image.Rectangle) bool
+
+	// Update the tracker, return new bounding box and a bool, if it lost it.
+	//
+	// https://docs.opencv.org/master/d0/d0a/classcv_1_1Tracker.html#a549159bd0553e6a8de356f3866df1f18
+	Update(image gocv.Mat) (image.Rectangle, bool)
+}
+
+//
+// available implementations in gocv are:
+//
+// TrackerMil;
+// TrackerBoosting;
+// TrackerMedianFlow;
+// TrackerTld;
+// TrackerKcf;
+// TrackerMosse;
+// TrackerCsrt;
+//
+
+// (private) implementation helpers for init and update (saves some boilerplate code)
+func tracker_Init(trk C.Tracker, img gocv.Mat, boundingBox image.Rectangle) bool {
+	cBox := C.struct_Rect{
+		x:      C.int(boundingBox.Min.X),
+		y:      C.int(boundingBox.Min.Y),
+		width:  C.int(boundingBox.Size().X),
+		height: C.int(boundingBox.Size().Y),
+	}
+
+	ret := C.Tracker_Init(trk, C.Mat(img.Ptr()), cBox)
+	return bool(ret)
+}
+
+func tracker_Update(trk C.Tracker, img gocv.Mat) (image.Rectangle, bool) {
+	cBox := C.struct_Rect{}
+
+	ret := C.Tracker_Update(trk, C.Mat(img.Ptr()), &cBox)
+
+	rect := image.Rect(int(cBox.x), int(cBox.y), int(cBox.x+cBox.width), int(cBox.y+cBox.height))
+	return rect, bool(ret)
+}
+
+//
+// Mil
+// The MIL algorithm trains a classifier in an online manner to separate the object from the background.
+// Multiple Instance Learning avoids the drift problem for a robust tracking.
+//
+// s.a: https://docs.opencv.org/master/d0/d26/classcv_1_1TrackerMIL.html
+type TrackerMil struct {
+	p C.TrackerMil
+}
+
+func NewTrackerMil() TrackerMil {
+	return TrackerMil{p: C.TrackerMil_Create()}
+}
+
+func (self *TrackerMil) Close() error {
+	C.TrackerMil_Close(self.p)
+	self.p = nil
+	return nil
+}
+
+func (self TrackerMil) Init(img gocv.Mat, boundingBox image.Rectangle) bool {
```

Comment:
Missing godoc here.

Reply:
hmm, isn't that only useless repetition ?

(those are all boring implementations of the same interface)

(but i'll have to look at the godoc output, again, let's see)

Reply:
Yes and no, it is a bit repetitive but it is useful to have documentation everywhere so that when you're looking for something, you dont have to look somewhere else first, even if that "somewhere" looks similar to where you were initially.